pkg/netflow/common/flow.go

```diff
@@ -0,0 +1,90 @@
+package common
+
+import (
+	"encoding/json"
+	"hash/fnv"
+	"strconv"
+)
+
+// Flow contains flow info used for aggregation
+// json annotations are used in AsJSONString() for debugging purpose
+type Flow struct {
+	Namespace    string   `json:"namespace"`
+	FlowType     FlowType `json:"type"`
+	SamplingRate uint64   `json:"sampling_rate"`
+	Direction    uint32   `json:"direction"`
+
+	// Exporter information
+	ExporterAddr string `json:"exporter_addr"`
+
+	// Flow time
+	StartTimestamp uint64 `json:"start_timestamp"` // in seconds
+	EndTimestamp   uint64 `json:"end_timestamp"`   // in seconds
+
+	// Size of the sampled packet
+	Bytes   uint64 `json:"bytes"`
+	Packets uint64 `json:"packets"`
+
+	// Source/destination addresses
+	SrcAddr string `json:"src_addr"` // FLOW KEY
+	DstAddr string `json:"dst_addr"` // FLOW KEY
+
+	// Layer 3 protocol (IPv4/IPv6/ARP/MPLS...)
+	EtherType uint32 `json:"ether_type"`
+
+	// Layer 4 protocol
+	IPProtocol uint32 `json:"ip_protocol"` // FLOW KEY
+
+	// Ports for UDP and TCP
+	SrcPort uint32 `json:"src_port"` // FLOW KEY
+	DstPort uint32 `json:"dst_port"` // FLOW KEY
+
+	// SNMP Interface Index
+	InputInterface  uint32 `json:"input_interface"` // FLOW KEY
+	OutputInterface uint32 `json:"output_interface"`
+
+	// Mac Address
+	SrcMac uint64 `json:"src_mac"`
+	DstMac uint64 `json:"dst_mac"`
+
+	// Mask
+	SrcMask uint32 `json:"src_mask"`
+	DstMask uint32 `json:"dst_mask"`
+
+	// Ethernet information
+	Tos uint32 `json:"tos"` // FLOW KEY
+
+	NextHop string `json:"next_hop"` // FLOW KEY
+}
+
+// AggregationHash return a hash used as aggregation key
+func (f *Flow) AggregationHash() string {
```

Comment:
Seems like in this method we do a bunch of conversion to strings which does not seem to be necessary and can end up being costly as it's on the hot path.

1. We probably don't need to cast from int -> string -> []byte, we can probably use a binary encoder to go from int -> []byte in a much quicker way.
2. I don't think we need to `FormatUint` at the end, can't we simply use `h.Sum64()` as the hash key?

Reply:
Agree, since it's in a hot path it can worth spending extra time to get more perf on this. Will check that.

Reply:
@FlorianVeaux 

1/ I refactor the AggregationHash function to not require int to string to bytes conversion.

See https://github.com/DataDog/datadog-agent/pull/10976/commits/78ffa87831e14024141c0ba3eb369e8063867af9#diff-1ac9946e13e183c729f0b8b9ec14f4ca61949baf94b3861a46a1fa0c951aa64aR59-R68

I'm also storing some fields like src IP, dst IP, next Hop IP as []byte instead of string in common.Flow to reduce storage space.
Those are  converted to string only when the payload is built.

2/ Good idea, fixed, uint64 will take less space and is likely faster to look up.